### PR TITLE
make math functions support complex arguments via anytype

### LINKED ISSUE #19207: Proposal: nuke the `std.math.complex` namespace and make math functions generic over `Complex`
These functions already uses `anytype` to dispatch to different datatype-specific implementations.

It would just be a matter ducktyping each function's `else` branch for presence of field names `re` and `im`.

The complex-specific implementations can stay in the `math/complex` directory or it can be just merged into the main files, both options should be fine but the main point is that there probably shouldn't be a public `complex` namespace exposed for confusion.

### PR DESCRIPTION
**Pull request by expikr**
An enhancement to automatically dispatch various math functions to the complex-valued version at comptime.

- makes various `math.*` functions accept both real arguments as well as structs which contain `re` and `im` fields
- add `math.ln` that takes real or complex args

closes #19207

### CODE REVIEW

**Comment by nektro** [comment]
the code doesnt pass `zig fmt`.

File: lib/std/math.zig [hunk 1]
```diff
@@ -272,25 +275,44 @@ pub const gcd = @import("math/gcd.zig").gcd;
...
 pub const lgamma = @import("math/gamma.zig").lgamma;
 
-/// Sine trigonometric function on a floating point number.
+/// Absolute value function on a real or complex number.
 /// Uses a dedicated hardware instruction when available.
-/// This is the same as calling the builtin @sin
+/// For real numbers, this is the same as calling the builtin @abs
+pub inline fn abs(value: anytype) complex.ScalarType(@TypeOf(value)) {
+    return switch (@TypeOf(value)) {
+        Complex(f32), Complex(f64) => complex.abs(value),
```
Comment (Snektron) [inline lib/std/math.zig:283]: Did you consider `Complex(f128)` and `Complex(comptime_float)`?

File: lib/std/math.zig [hunk 2]
```diff
@@ -335,7 +370,7 @@ pub inline fn exp2(value: anytype) @TypeOf(value) {
     return @exp2(value);
 }
 
-pub const complex = @import("math/complex.zig");
+const complex = @import("math/complex.zig");
```
Comment (Snektron) [inline lib/std/math.zig:373]: This seems likely to break a bunch of code. Im not sure that its necessary to hide the `complex` namespace

File: lib/std/math.zig [hunk 3]
```diff
@@ -426,7 +484,6 @@ pub const min = @compileError("deprecated; use @min instead");
 pub const max = @compileError("deprecated; use @max instead");
 pub const min3 = @compileError("deprecated; use @min instead");
 pub const max3 = @compileError("deprecated; use @max instead");
-pub const ln = @compileError("deprecated; use @log instead");
```
Comment (Snektron) [inline lib/std/math.zig:429]: I think this should just say `use @log or std.math.log instead`

File: lib/std/math.zig (hunk 1 above)
Comment (Vexu) [reply lib/std/math.zig:283]: You didn't mark `isComplex` as inline.

File: lib/std/math.zig [hunk 4]
```diff
@@ -397,12 +415,34 @@ test {
     _ = gamma;
     _ = lgamma;
 
-    _ = complex;
     _ = Complex;
 
     _ = big;
 }
 
+test "Complex inputs" {
```
Comment (Vexu) [inline lib/std/math.zig:423]: It'd be nice if this tested anything like the functions returning the correct type or a sensible value.

**Review by Vexu — APPROVED** [review]

**Pull request closed without merging**